File: fe/fe-core/src/test/java/org/apache/doris/common/plugin/CloudPluginDownloaderTest.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.common.plugin;

import org.apache.doris.cloud.proto.Cloud;
import org.apache.doris.common.plugin.CloudPluginDownloader.PluginType;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

public class CloudPluginDownloaderTest {

    @TempDir
    Path tempDir;

    @Test
    void testValidateInput() {
        Assertions.assertDoesNotThrow(() -> {
            CloudPluginDownloader.validateInput(PluginType.JDBC_DRIVERS, "mysql.jar");
            CloudPluginDownloader.validateInput(PluginType.JAVA_UDF, "nested/my_udf-1.0@prod.jar");
        });

        Assertions.assertEquals("Plugin name cannot be empty",
                Assertions.assertThrows(IllegalArgumentException.class,
                        () -> CloudPluginDownloader.validateInput(PluginType.JDBC_DRIVERS, ""))
                        .getMessage());
        Assertions.assertEquals("Plugin name cannot be empty",
                Assertions.assertThrows(IllegalArgumentException.class,
                        () -> CloudPluginDownloader.validateInput(PluginType.JDBC_DRIVERS, null))
                        .getMessage());
        Assertions.assertThrows(UnsupportedOperationException.class,
                () -> CloudPluginDownloader.validateInput(PluginType.CONNECTORS, "test.jar"));

        for (String invalidName : new String[] {
                "../driver.jar", "nested/../../driver.jar", "nested/../driver.jar",
                "./driver.jar", "/driver.jar", "driver?.jar", "driver#v1.jar",
                "driver%20v1.jar", "nested\\driver.jar", "driver.txt"
        }) {
            Assertions.assertThrows(IllegalArgumentException.class,
                    () -> CloudPluginDownloader.validateInput(PluginType.JDBC_DRIVERS, invalidName),
                    invalidName);
        }
    }

    @Test
    void testDownloadRejectsTargetOutsidePluginDirectoryBeforeCloudAccess() {
        Assertions.assertThrows(IllegalArgumentException.class,
                () -> CloudPluginDownloader.downloadFromCloud(
                        PluginType.JDBC_DRIVERS, "driver.jar", tempDir.resolve("driver.jar").toString()));
    }

    @Test
    void testSelectLatestCloudStorageInfo() {
        Cloud.ObjectStoreInfoPB oldInfo = objectStoreInfo("old-bucket", "old-prefix");
        Cloud.ObjectStoreInfoPB latestInfo = objectStoreInfo("latest-bucket", "latest-prefix");
        Cloud.GetObjStoreInfoResponse response = responseBuilder()
                .addObjInfo(oldInfo)
                .addObjInfo(latestInfo)
                .build();

        Assertions.assertEquals(latestInfo, CloudPluginDownloader.selectCloudStorageInfo(response));
    }

    @Test
    void testRejectInvalidCloudStorageResponses() {
        Cloud.GetObjStoreInfoResponse failed = Cloud.GetObjStoreInfoResponse.newBuilder()
                .setStatus(Cloud.MetaServiceResponseStatus.newBuilder()
                        .setCode(Cloud.MetaServiceCode.INVALID_ARGUMENT)
                        .setMsg("test error"))
                .build();
        Assertions.assertTrue(Assertions.assertThrows(RuntimeException.class,
                () -> CloudPluginDownloader.selectCloudStorageInfo(failed)).getMessage()
                .contains("Failed to get storage info"));

        Assertions.assertTrue(Assertions.assertThrows(RuntimeException.class,
                () -> CloudPluginDownloader.selectCloudStorageInfo(responseBuilder().build())).getMessage()
                .contains("Only SaaS cloud storage is supported"));

        Cloud.GetObjStoreInfoResponse storageVault = responseBuilder()
                .setEnableStorageVault(true)
                .addObjInfo(objectStoreInfo("vault-bucket", "vault-prefix"))
                .build();
        Assertions.assertTrue(Assertions.assertThrows(RuntimeException.class,
                () -> CloudPluginDownloader.selectCloudStorageInfo(storageVault)).getMessage()
                .contains("legacy SaaS mode"));
    }

    @Test
    void testBuildS3Path() {
        Cloud.ObjectStoreInfoPB withPrefix = objectStoreInfo("test-bucket", "test-prefix");
        Assertions.assertEquals("s3://test-bucket/test-prefix/plugins/jdbc_drivers/mysql.jar",
                CloudPluginDownloader.buildS3Path(withPrefix, PluginType.JDBC_DRIVERS, "mysql.jar"));
        Assertions.assertEquals("s3://test-bucket/test-prefix/plugins/java_udf/nested/my_udf.jar",
                CloudPluginDownloader.buildS3Path(withPrefix, PluginType.JAVA_UDF, "nested/my_udf.jar"));

        Cloud.ObjectStoreInfoPB withoutPrefix = objectStoreInfo("test-bucket", null);
        Assertions.assertEquals("s3://test-bucket/plugins/java_udf/test-udf@v1.0.jar",
                CloudPluginDownloader.buildS3Path(withoutPrefix, PluginType.JAVA_UDF,
                        "test-udf@v1.0.jar"));
    }

    @Test
    void testDownloadPublishesCompleteFileAtomically() throws Exception {
        Path target = tempDir.resolve("driver.jar");
        Files.write(target, "old".getBytes(StandardCharsets.UTF_8));

        String result = CloudPluginDownloader.downloadToLocal(target,
                () -> new ByteArrayInputStream("new-driver".getBytes(StandardCharsets.UTF_8)));

        Assertions.assertEquals(target.toAbsolutePath().toString(), result);
        Assertions.assertEquals("new-driver", new String(Files.readAllBytes(target), StandardCharsets.UTF_8));
        assertOnlyTargetRemains(target);
    }

    @Test
    void testDownloadFailurePreservesExistingFileAndClosesStream() throws Exception {
        Path target = tempDir.resolve("driver.jar");
        Files.write(target, "old-driver".getBytes(StandardCharsets.UTF_8));
        AtomicBoolean closed = new AtomicBoolean();

        Assertions.assertThrows(IOException.class, () -> CloudPluginDownloader.downloadToLocal(target,
                () -> failingStream(closed)));

        Assertions.assertTrue(closed.get());
        Assertions.assertEquals("old-driver", new String(Files.readAllBytes(target), StandardCharsets.UTF_8));
        assertOnlyTargetRemains(target);
    }

    @Test
    void testConcurrentDownloadsOfSameTargetAreSerialized() throws Exception {
        Path target = tempDir.resolve("driver.jar");
        CountDownLatch firstStarted = new CountDownLatch(1);
        CountDownLatch releaseFirst = new CountDownLatch(1);
        CountDownLatch secondStarted = new CountDownLatch(1);
        ExecutorService executor = Executors.newFixedThreadPool(2);
        try {
            Future<String> first = executor.submit(() -> CloudPluginDownloader.downloadToLocal(target, () -> {
                firstStarted.countDown();
                releaseFirst.await();
                return new ByteArrayInputStream("first".getBytes(StandardCharsets.UTF_8));
            }));
            Assertions.assertTrue(firstStarted.await(5, TimeUnit.SECONDS));

            Future<String> second = executor.submit(() -> CloudPluginDownloader.downloadToLocal(target, () -> {
                secondStarted.countDown();
                return new ByteArrayInputStream("second".getBytes(StandardCharsets.UTF_8));
            }));
            Assertions.assertFalse(secondStarted.await(200, TimeUnit.MILLISECONDS));

            releaseFirst.countDown();
            first.get(5, TimeUnit.SECONDS);
            second.get(5, TimeUnit.SECONDS);
            Assertions.assertTrue(secondStarted.await(5, TimeUnit.SECONDS));
            Assertions.assertEquals("second", new String(Files.readAllBytes(target), StandardCharsets.UTF_8));
            assertOnlyTargetRemains(target);
        } finally {
            releaseFirst.countDown();
            executor.shutdownNow();
        }
    }

    @Test
    void testDownloadsOfDifferentTargetsCanRunConcurrently() throws Exception {
        Path firstTarget = tempDir.resolve("first.jar");
        Path secondTarget = tempDir.resolve("second.jar");
        CountDownLatch firstStarted = new CountDownLatch(1);
        CountDownLatch releaseFirst = new CountDownLatch(1);
        CountDownLatch secondStarted = new CountDownLatch(1);
        ExecutorService executor = Executors.newFixedThreadPool(2);
        try {
            Future<String> first = executor.submit(() -> CloudPluginDownloader.downloadToLocal(
                    firstTarget, () -> {
                        firstStarted.countDown();
                        releaseFirst.await();
                        return new ByteArrayInputStream("first".getBytes(StandardCharsets.UTF_8));
                    }));
            Assertions.assertTrue(firstStarted.await(5, TimeUnit.SECONDS));

            Future<String> second = executor.submit(() -> CloudPluginDownloader.downloadToLocal(
                    secondTarget, () -> {
                        secondStarted.countDown();
                        return new ByteArrayInputStream("second".getBytes(StandardCharsets.UTF_8));
                    }));
            Assertions.assertTrue(secondStarted.await(5, TimeUnit.SECONDS));
            second.get(5, TimeUnit.SECONDS);

            releaseFirst.countDown();
            first.get(5, TimeUnit.SECONDS);
            Assertions.assertEquals("first",
                    new String(Files.readAllBytes(firstTarget), StandardCharsets.UTF_8));
            Assertions.assertEquals("second",
                    new String(Files.readAllBytes(secondTarget), StandardCharsets.UTF_8));
        } finally {
            releaseFirst.countDown();
            executor.shutdownNow();
        }
    }

    private static Cloud.GetObjStoreInfoResponse.Builder responseBuilder() {
        return Cloud.GetObjStoreInfoResponse.newBuilder()
                .setStatus(Cloud.MetaServiceResponseStatus.newBuilder()
                        .setCode(Cloud.MetaServiceCode.OK));
    }

    private static Cloud.ObjectStoreInfoPB objectStoreInfo(String bucket, String prefix) {
        Cloud.ObjectStoreInfoPB.Builder builder = Cloud.ObjectStoreInfoPB.newBuilder()
                .setProvider(Cloud.ObjectStoreInfoPB.Provider.S3)
                .setBucket(bucket);
        if (prefix != null) {
            builder.setPrefix(prefix);
        }
        return builder.build();
    }

    private static InputStream failingStream(AtomicBoolean closed) {
        return new InputStream() {
            private boolean firstRead = true;

            @Override
            public int read(byte[] bytes, int offset, int length) throws IOException {
                if (firstRead) {
                    firstRead = false;
                    bytes[offset] = 'x';
                    return 1;
                }
                throw new IOException("injected download failure");
            }

            @Override
            public int read() throws IOException {
                throw new IOException("injected download failure");
            }

            @Override
            public void close() {
                closed.set(true);
            }
        };
    }

    private static void assertOnlyTargetRemains(Path target) throws IOException {
        try (java.util.stream.Stream<Path> files = Files.list(target.getParent())) {
            Assertions.assertArrayEquals(new Path[] {target}, files.toArray(Path[]::new));
        }
    }
}

File: be/test/runtime/plugin/cloud_plugin_downloader_test.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "runtime/plugin/cloud_plugin_downloader.h"

#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include <filesystem>
#include <fstream>

#include "cloud/cloud_storage_engine.h"
#include "olap/storage_engine.h"
#include "runtime/exec_env.h"

namespace doris {

class CloudPluginDownloaderTest : public ::testing::Test {
protected:
    void SetUp() override { downloader = std::make_unique<CloudPluginDownloader>(); }

    void TearDown() override {
        downloader.reset();
        ExecEnv::GetInstance()->set_storage_engine(nullptr);
    }

    void SetupRegularStorageEngine() {
        ExecEnv::GetInstance()->set_storage_engine(
                std::make_unique<StorageEngine>(EngineOptions()));
    }

    void SetupCloudStorageEngine() {
        ExecEnv::GetInstance()->set_storage_engine(
                std::make_unique<CloudStorageEngine>(EngineOptions()));
    }

    std::unique_ptr<CloudPluginDownloader> downloader;
};

// ============== Core Business Logic Tests ==============

TEST_F(CloudPluginDownloaderTest, TestBuildPluginPath) {
    std::string path;
    Status status = downloader->_build_plugin_path(CloudPluginDownloader::PluginType::JDBC_DRIVERS,
                                                   "mysql-connector.jar", &path);
    EXPECT_TRUE(status.ok());
    EXPECT_EQ("plugins/jdbc_drivers/mysql-connector.jar", path);

    status = downloader->_build_plugin_path(CloudPluginDownloader::PluginType::JAVA_UDF,
                                            "my-udf.jar", &path);
    EXPECT_TRUE(status.ok());
    EXPECT_EQ("plugins/java_udf/my-udf.jar", path);
}

TEST_F(CloudPluginDownloaderTest, TestBuildPluginPathEdgeCases) {
    std::string path;
    Status status = downloader->_build_plugin_path(CloudPluginDownloader::PluginType::JDBC_DRIVERS,
                                                   "test-file_v1.2.jar", &path);
    EXPECT_TRUE(status.ok());
    EXPECT_EQ("plugins/jdbc_drivers/test-file_v1.2.jar", path);

    status = downloader->_build_plugin_path(CloudPluginDownloader::PluginType::JAVA_UDF,
                                            "sub/dir/file.jar", &path);
    EXPECT_TRUE(status.ok());
    EXPECT_EQ("plugins/java_udf/sub/dir/file.jar", path);

    std::string long_name(100, 'a');
    long_name += ".jar";
    status = downloader->_build_plugin_path(CloudPluginDownloader::PluginType::JDBC_DRIVERS,
                                            long_name, &path);
    EXPECT_TRUE(status.ok());
    std::string expected = "plugins/jdbc_drivers/" + long_name;
    EXPECT_EQ(expected, path);
}

TEST_F(CloudPluginDownloaderTest, TestBuildPluginPathUnsupportedTypes) {
    std::string path;
    Status status = downloader->_build_plugin_path(CloudPluginDownloader::PluginType::CONNECTORS,
                                                   "kafka-connector.jar", &path);
    EXPECT_FALSE(status.ok());
    EXPECT_EQ(status.code(), ErrorCode::INVALID_ARGUMENT);
    EXPECT_THAT(status.msg(), testing::HasSubstr("Unsupported plugin type"));

    status = downloader->_build_plugin_path(CloudPluginDownloader::PluginType::HADOOP_CONF,
                                            "core-site.xml", &path);
    EXPECT_FALSE(status.ok());
    EXPECT_EQ(status.code(), ErrorCode::INVALID_ARGUMENT);
    EXPECT_THAT(status.msg(), testing::HasSubstr("Unsupported plugin type"));
}

TEST_F(CloudPluginDownloaderTest, TestBuildPluginPathInvalidType) {
    CloudPluginDownloader::PluginType invalid_type =
            static_cast<CloudPluginDownloader::PluginType>(999);
    std::string path;
    Status status = downloader->_build_plugin_path(invalid_type, "test.jar", &path);
    EXPECT_FALSE(status.ok());
    EXPECT_EQ(status.code(), ErrorCode::INVALID_ARGUMENT);
    EXPECT_THAT(status.msg(), testing::HasSubstr("Unsupported plugin type"));
}

// ============== Storage Engine Integration Tests ==============

TEST_F(CloudPluginDownloaderTest, TestDownloadFromCloudEmptyName) {
    std::string result_path;
    Status status = CloudPluginDownloader::download_from_cloud(
            CloudPluginDownloader::PluginType::JDBC_DRIVERS, "", "/tmp/test.jar", &result_path);

    EXPECT_FALSE(status.ok());
    EXPECT_EQ(status.code(), ErrorCode::INVALID_ARGUMENT);
    EXPECT_EQ("Plugin name cannot be empty", status.msg());
}

TEST_F(CloudPluginDownloaderTest, TestDownloadFromCloudRegularStorageEngine) {
    SetupRegularStorageEngine();

    std::string result_path;
    Status status = CloudPluginDownloader::download_from_cloud(
            CloudPluginDownloader::PluginType::JDBC_DRIVERS, "mysql.jar", "/tmp/test.jar",
            &result_path);

    EXPECT_FALSE(status.ok());
    EXPECT_EQ(status.code(), ErrorCode::NOT_FOUND);
    EXPECT_TRUE(status.to_string().find("CloudStorageEngine not found") != std::string::npos);
}

TEST_F(CloudPluginDownloaderTest, TestGetCloudFilesystemNonCloudEnvironment) {
    SetupRegularStorageEngine();

    io::RemoteFileSystemSPtr filesystem;
    Status status = downloader->_get_cloud_filesystem(&filesystem);

    EXPECT_FALSE(status.ok());
    EXPECT_EQ(status.code(), ErrorCode::NOT_FOUND);
    EXPECT_TRUE(status.to_string().find("CloudStorageEngine not found") != std::string::npos);
}

TEST_F(CloudPluginDownloaderTest, TestGetCloudFilesystemNoStorageEngine) {
    SetupRegularStorageEngine();

    io::RemoteFileSystemSPtr filesystem;
    Status status = downloader->_get_cloud_filesystem(&filesystem);

    EXPECT_FALSE(status.ok());
    EXPECT_EQ(status.code(), ErrorCode::NOT_FOUND);
}

TEST_F(CloudPluginDownloaderTest, TestGetCloudFilesystemCloudEnvironment) {
    SetupCloudStorageEngine();
    io::RemoteFileSystemSPtr filesystem;
    Status status = downloader->_get_cloud_filesystem(&filesystem);
    if (!status.ok()) {
        EXPECT_EQ(status.code(), ErrorCode::NOT_FOUND);
        EXPECT_TRUE(status.to_string().find("No latest filesystem available") != std::string::npos);
    }
}

TEST_F(CloudPluginDownloaderTest, TestDownloadFromCloudCloudStorageEngine) {
    SetupCloudStorageEngine();
    std::string result_path;
    Status status = CloudPluginDownloader::download_from_cloud(
            CloudPluginDownloader::PluginType::JDBC_DRIVERS, "mysql.jar", "/tmp/test.jar",
            &result_path);
    EXPECT_FALSE(status.ok());
}

// ============== File System Tests ==============

TEST_F(CloudPluginDownloaderTest, TestPrepareLocalPathSuccess) {
    std::string test_path = "/tmp/test_new_file.jar";
    Status status = downloader->_prepare_local_path(test_path);
    EXPECT_TRUE(status.ok());
}

TEST_F(CloudPluginDownloaderTest, TestPrepareLocalPathWithExistingFile) {
    std::string existing_file = "/tmp/existing_file.jar";
    std::ofstream file(existing_file);
    file << "existing content";
    file.close();
    EXPECT_TRUE(std::filesystem::exists(existing_file));

    Status status = downloader->_prepare_local_path(existing_file);
    EXPECT_TRUE(status.ok());
    EXPECT_FALSE(std::filesystem::exists(existing_file));
}

TEST_F(CloudPluginDownloaderTest, TestPrepareLocalPathWithNestedDirectory) {
    std::string nested_path = "/tmp/test_dir/nested/test.jar";
    Status status = downloader->_prepare_local_path(nested_path);
    EXPECT_TRUE(status.ok());
    EXPECT_TRUE(std::filesystem::exists("/tmp/test_dir/nested"));
    std::filesystem::remove_all("/tmp/test_dir");
}

TEST_F(CloudPluginDownloaderTest, TestPrepareLocalPathRootDirectory) {
    std::string root_path = "/test_root.jar";
    Status status = downloader->_prepare_local_path(root_path);
    EXPECT_TRUE(status.ok());
    std::filesystem::remove(root_path);
}

// ============== Consistency Tests ==============

TEST_F(CloudPluginDownloaderTest, TestAllTypePathCombinations) {
    struct TestCase {
        CloudPluginDownloader::PluginType type;
        std::string name;
        std::string expected;
    };

    std::vector<TestCase> test_cases = {
            {CloudPluginDownloader::PluginType::JDBC_DRIVERS, "driver.jar",
             "plugins/jdbc_drivers/driver.jar"},
            {CloudPluginDownloader::PluginType::JAVA_UDF, "udf.jar", "plugins/java_udf/udf.jar"},
    };

    for (const auto& test_case : test_cases) {
        std::string result;
        Status status = downloader->_build_plugin_path(test_case.type, test_case.name, &result);
        EXPECT_TRUE(status.ok()) << "Failed for type: " << static_cast<int>(test_case.type)
                                 << ", name: " << test_case.name << ", error: " << status.msg();
        EXPECT_EQ(test_case.expected, result)
                << "Failed for type: " << static_cast<int>(test_case.type)
                << ", name: " << test_case.name;
    }
}

} // namespace doris